libGraphite/spriteworld/rleX.cpp

```diff
@@ -326,53 +158,67 @@ auto graphite::spriteworld::rleX::decode(data::reader &reader) -> void
-
-    while (!completed_last_frame) {
-        opcode_lut[reader.read_byte()](
-            this,
-            m_surface,
-            reader,
-            frame,
-            current_offset,
-            right_bound,
-            pitch,
-            current_frame,
-            m_frame_count,
-            completed_last_frame,
-            yuv
-        );
+    std::uint32_t pitch = m_surface.size().width - m_frame_size.width;
+
+    auto rect = frame_rect(0);
+    auto raw = surface().raw().get<std::uint8_t *>();
+
+    
+    for (auto frame = 0; frame < m_frame_count; ++frame) {
+        rect = this->frame_rect(frame);
+        for (auto ch = 0; ch < 4; ++ch) {
+            offset = (rect.origin.y * m_surface.size().width) + rect.origin.x;
+            right_bound = offset + m_frame_size.width;
+            auto pack_size = reader.read_long();
+            auto pack = reader.read_data(pack_size).get<std::uint8_t *>();
+            auto pack_offset = 0;
+
+            while (pack_offset < pack_size) {
+                auto run = static_cast<std::uint32_t>(pack[pack_offset++]);
+                if (run < 0x80) {
+                    // Copy bytes
+                    run++;
+                    if (pack_offset + run > pack_size) {
+                        throw std::runtime_error("Insufficient data for rlëX resource: " + std::to_string(m_id) + ", " + m_name);
+                    }
+                    for (auto i = 0; i < run; ++i) {
+                        raw[offset*4 + ch] = pack[pack_offset++];
+                        if (++offset >= right_bound) {
+                            offset += pitch;
+                            right_bound = offset + m_frame_size.width;
+                        }
+                    }
+                }
+                else {
+                    // Repeat single byte
+                    run ^= 0x80;
+                    if (run >= 0x70) {
+                        // 2 byte count
+                        if (pack_offset == pack_size) {
+                            throw std::runtime_error("Insufficient data for rlëX resource: " + std::to_string(m_id) + ", " + m_name);
+                        }
+                        // Combine 4 low bits with next byte
+                        run = (run & 0x0F) << 8 | pack[pack_offset++];
+                    }
+                    run++;
+                    if (pack_offset == pack_size) {
+                        throw std::runtime_error("Insufficient data for rlëX resource: " + std::to_string(m_id) + ", " + m_name);
+                    }
+                    auto value = pack[pack_offset++];
+                    for (auto i = 0; i < run; ++i) {
+                        raw[offset*4 + ch] = value;
+                        if (++offset >= right_bound) {
+                            offset += pitch;
+                            right_bound = offset + m_frame_size.width;
+                        }
+                    }
+                }
+            }
+        }
     }
```

Comment:
Again,  I'm going to need to verify the performance of this. I moved to a look up table to eliminate the excessive cpu cycles resulting from conditionals. Given that the goal of this PR is to further compress and reduce the size of the `rlëX` resource, it is invariably going to result in degraded performance, as more work will be required to decompress.

🤔